### ISSUE TITLE
Can't understand how to use option<T>.MatchAsync

### ISSUE THREAD

**Issue by MrYossu** [louthy/language-ext#854]
I have the following...

```c#
    Option<Investor> Investor { get; set; }
```
...and I want use it to populate a `RenewalViewModel`. Up until now, my code looked like this...

```c#
    public async Task<RenewalViewModel> NewRenewalViewModel() {
      var jim = await GetJimData();
      return Investor.Match(
       inv => {
          return new RenewalViewModel {
            // Set properties here, based on the investor and jim
          };
        },
      // investor is `None`, so return an empty view model
       new RenewalViewModel());
```

However, I now have a need to do some async stuff inside the `Some` branch of `Match`.

I tried the following...

```c#
    public async Task<RenewalViewModel> NewRenewalViewModel() {
      var jim = await GetJimData();
      return Investor.Match(
       async inv => {
          var jam = await GetJamData();
          return new RenewalViewModel {
            // Set properties here, based on the investor, jim and jam
          };
        },
       new RenewalViewModel());
```

...but got a compiler error on the 5th line "_Anonymous function converted to a void returning delegate cannot return a value_"

I tried using `MatchAsync` instead, but that gave a compiler error in the 3rd line "_cannot convert from 'RenewalViewModel' to 'System.Func<RenewalViewModel>'_"

Anyone able to advise how I do this? Thanks

**Comment by TysonMN** [comment]
Try replacing `Investor.Match(` with `Investor.ToAsync().Match(`.

**Comment by MrYossu** [comment]
Nope, still get that first error on line 5.

Thanks anyway, any other ideas?

**Comment by TysonMN** [comment]
Can you include a completely minimal working example?  Ideally a GitHub repo.

**Comment by MrYossu** [comment]
Working on it, but it's not easy!

**Comment by MrYossu** [comment]
Right, I think I have a sample ready. Please take a look at [this GitHub repo](https://github.com/MrYossu/AsyncOptionMatch).

The scenario is that I have investors, each of whom can have many terms (usually resubscribe every year, so anew term per year). In each term, they have a number of shares. 

The `InvestorController` class simulates what I'm trying to do. There is a method that sets up the `Option<Investor>`, and another that returns a `RenewalViewModel`. It's that second method that is causing me the problem.

The method contains an `async` call inside the `Some` branch of `Match`...

    List<Share> shares = await _appDbContext.Shares();

However, this causes the compiler error. If I remove that line, and the `async` qualifier at the beginning of the lambda on line 19, it compiles. However, I need to make this `async` call inside the lambda.

I hope that's clearer. Thanks again

**Comment by TysonMN** [comment]
Unfortunately, it is not clear to me.  What do you want to do with that `share` variable?  It is currently unread.

**Comment by MrYossu** [comment]
Sorry, I was too hasty. The `shares` variable is read form the database (which is why it needs to be async) and then will be used when setting one of the properties on the `RenewalViewModel`.

I just updated the repo code to use it.

In reality there would be some extra stuff going on, such as filtering the shares, but the principle would be the same.

Thanks

**Comment by TysonMN** [comment]
> > Try replacing `Investor.Match(` with `Investor.ToAsync().Match(`.
> 
> Nope, still get that first error on line 5.
> 
> Thanks anyway, any other ideas?

I didn't mean that the _only_ change was adding `ToAsync`.  That was just the first change to get you in the right direction.  Sorry if that was unclear.

After that, I prefer to use `MapAsync` (instead of some `Match` function) and finish with `IfNone`.

```csharp
public Task SetUpInvestor(int id) {
  _investor = _appDbContext.Investors.FirstOrDefault(i => i.Id == id);
  return Task.CompletedTask;
}

public Task<RenewalViewModel> NewRenewalViewModel() =>
  _investor.ToAsync().MapAsync(
   async inv => {
      Term latestTerm = inv.Terms.OrderByDescending(ila => ila.End).First();
      List<Share> shares = await _appDbContext.Shares(); // In reality we would be doing some filtering here. Omitted for clarity
      return new RenewalViewModel {
        Start = latestTerm.End.AddDays(1),
        End = latestTerm.End.AddYears(1),
        Shares = latestTerm.Shares.Select(s => new Share {
          Id = s.Id,
          Quantity = s.Quantity,
          CompanyId = s.CompanyId,
        }).ToList(),
        OtherShares = shares,
      };
    })
    .IfNone(new RenewalViewModel());
```

**Comment by MrYossu** [comment]
Thanks for that, I see what you mean now. Never thought of doing it that way.

Thanks again.

**Comment by ode2code95** [comment]
@TysonMN, `MapAsync` actually has an overload that accepts an option and internally calls `ToAsync` for you.
